alvr/sockets/src/stream_socket/mod.rs

```diff
@@ -0,0 +1,397 @@
+
+// Performance analysis:
+// We want to minimize the transmission time for various sizes of packets.
+// The current code locks the write socket *per shard* and not *per packet*. This leds to the best
+// performance outcome given that the possible packets can be either very small (one shard) or very
+// large (hundreds/thousands of shards, for video). if we don't allow interleaving shards, a very
+// small packet will need to wait a long time before getting received if there was an ongoing
+// transmission of a big packet before. If we allow interleaving shards, small packets can be
+// transmitted quicker, with only minimal latency increase for the ongoing transmission of the big
+// packet.
+// Note: We can't clone the underlying socket for each StreamSender and the mutex around the socket
+// cannot be removed. This is because we need to make sure at least shards are written whole.
+
+mod tcp;
+mod udp;
+
+use alvr_common::{
+    AnyhowToCon, ConResult, HandleTryAgain, ToCon, anyhow::Result, parking_lot::Mutex,
+};
+use alvr_session::{DscpTos, SocketBufferSize, SocketProtocol};
+use bincode::config;
+use serde::{Serialize, de::DeserializeOwned};
+use std::{
+    cmp::Ordering,
+    collections::HashMap,
+    marker::PhantomData,
+    mem,
+    net::{IpAddr, TcpListener, UdpSocket},
+    ops::Range,
+    sync::{Arc, mpsc},
+    time::Duration,
+};
+
+trait MultiplexedSocketWriter {
+    // Note: consts are not trait-safe, we require a method
+    fn payload_offset(&self) -> usize;
+
+    fn send(&mut self, stream_id: u16, packet_index: u32, buffer: &mut Vec<u8>) -> Result<()>;
+}
+
+struct ReconstructedPacket {
+    index: u32,
+    buffer: Vec<u8>,
+}
+
+struct StreamRecvQueues {
+    used_buffer_sender: mpsc::Sender<Vec<u8>>,
+    used_buffer_receiver: mpsc::Receiver<Vec<u8>>,
+    packet_queue: mpsc::Sender<ReconstructedPacket>,
+}
+
+trait MultiplexedSocketReader {
+    fn payload_offset(&self) -> usize;
+
+    fn recv(&mut self, stream_queues: &HashMap<u16, StreamRecvQueues>) -> ConResult;
+}
+
+/// Memory buffer that contains a hidden prefix
+#[derive(Default)]
+pub struct Buffer<H = ()> {
+    inner: Vec<u8>,
+    raw_payload_offset: usize, // this corresponds to prefix + header
+    _phantom: PhantomData<H>,
+}
+
+impl<H> Buffer<H> {
+    /// Length of payload (without prefix)
+    #[must_use]
+    pub fn len(&self) -> usize {
+        self.inner.len() - self.raw_payload_offset
+    }
+
+    #[must_use]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Get the whole payload of the buffer
+    #[must_use]
+    pub fn get(&self) -> &[u8] {
+        &self.inner[self.raw_payload_offset..]
+    }
+
+    /// If the range is outside the valid range, new space will be allocated
+    /// NB: the offset parameter is applied on top of the internal offset of the buffer
+    #[must_use]
+    pub fn get_range_mut(&mut self, range: Range<usize>) -> &mut [u8] {
```

Comment:
Since we're reusing buffers I think this could really use a comment that it starts off with zero non-prefix memory so calling this doesn't run risk to have unused memory at the end. I got confused by that for a while.

Reply:
alright, fair

Reply:
The comment reads like it's talking about the backing memory, not the actual payload size.
I think something like this would be better:
> Newly acquired buffers will have no payload. If the provided range is outside the current payload size, the payload will be zero-extended to exactly fit the provided range.